src/proxy.cpp

```diff
@@ -105,6 +129,258 @@ int forward(
+        }
+    }
+
+
+    int i;
+    bool request_processed, reply_processed;
+
+
+    while (state != terminated) {
+
+        //  Blocking wait initially only for 'ZMQ_POLLIN' - 'poller_wait' points to 'poller_in'.
+        //  If one of receiving end's queue is full ('ZMQ_POLLOUT' not available),
+        //  'poller_wait' is pointed to 'poller_receive_blocked', 'poller_send_blocked' or 'poller_both_blocked'.
+        rc = poller_wait->wait (events, 3, -1);
+        if (rc < 0 && errno == ETIMEDOUT)
+            rc = 0;
+        CHECK_RC_EXIT_ON_FAILURE ();
+
+        //  Some of events waited for by 'poller_wait' have arrived, now poll for everything without blocking.
+        rc = poller_all->wait (events, 3, 0);
+        if (rc < 0 && errno == ETIMEDOUT)
+            rc = 0;
+        CHECK_RC_EXIT_ON_FAILURE ();
+
+        //  Process events.
+        for (i = 0; i < rc; i++) {
+            if (events [i].socket == frontend_) {
+                frontend_in = (events [i].events & ZMQ_POLLIN) != 0;
+                frontend_out = (events [i].events & ZMQ_POLLOUT) != 0;
+            } else
+                //  This 'if' needs to be after check for 'frontend_' in order never
+                //  to be reached in case frontend_==backend_, so we ensure backend_in=false in that case.
+                if (events [i].socket == backend_) {
+                    backend_in = (events [i].events & ZMQ_POLLIN) != 0;
+                    backend_out = (events [i].events & ZMQ_POLLOUT) != 0;
+                } else
+                    if (events [i].socket == control_)
+                        control_in = (events [i].events & ZMQ_POLLIN) != 0;
+        }
+
+
+        //  Process a control command if any.
+        if (control_in) {
+            rc = control_->recv (&msg, 0);
+            CHECK_RC_EXIT_ON_FAILURE ();
+            rc = control_->getsockopt (ZMQ_RCVMORE, &more, &moresz);
+            if (unlikely (rc < 0) || more) {
+                PROXY_CLEANUP ();
+                return close_and_return (&msg, -1);
+            }
+
+            //  Copy message to capture socket if any.
+            rc = capture (capture_, msg);
+            CHECK_RC_EXIT_ON_FAILURE ();
+
+            if (msg.size () == 5 && memcmp (msg.data (), "PAUSE", 5) == 0) {
+                state = paused;
+                poller_wait = poller_control;
+            } else
+                if (msg.size () == 6 && memcmp (msg.data (), "RESUME", 6) == 0) {
+                    state = active;
+                    poller_wait = poller_in;
+                } else
+                    if (msg.size () == 9 && memcmp (msg.data (), "TERMINATE", 9) == 0)
+                        state = terminated;
+                    else {
+                        //  This is an API error, we assert
+                        puts ("E: invalid command sent to proxy");
+                        zmq_assert (false);
+                    }
+                    control_in = false;
```

Comment:
@bjovke this snippet causes a build failure with GCC 6:

```
[  133s] src/proxy.cpp: In function 'int zmq::proxy(zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*, zmq::socket_base_t*)':
[  133s] src/proxy.cpp:322:19: error: this 'else' clause does not guard... [-Werror=misleading-indentation]
[  133s]                  } else
[  133s]                    ^~~~
[  133s] src/proxy.cpp:330:21: note: ...this statement, but the latter is misleadingly indented as if it is guarded by the 'else'
[  133s]                      control_in = false;
[  133s]                      ^~~~~~~~~~
```

The control_in assignment is aligned as if it was under the "else" block at line 307, but it is not, it's out of this if-else chain.
If that's intended please adjust the indentation, otherwise please fix the braces.